File: identity_agent_ui/rust/src/pqc/hybrid_inception.rs
//! Build hybrid PQC KERI-conformant hybrid icp events — Rust bridge (keri 1.1.17).

use serde::Serialize;
use serde_json::{json, Value};

use super::cesr::*;
use super::keri_serialize::{makify_icp_wire, versify, AnchorSeal, IcpWire, SAID_DUMMY_LEN};

#[derive(Debug, Clone)]
pub struct HybridKeyMaterial {
    pub ed25519_signing_raw: Vec<u8>,
    pub mldsa65_signing_raw: Vec<u8>,
    pub x25519_agreement_raw: Vec<u8>,
    pub mlkem768_encap_raw: Vec<u8>,
    pub next_ed25519_signing_raw: Vec<u8>,
    pub next_mldsa65_signing_raw: Vec<u8>,
}

pub fn synthetic_hybrid_key_material(seed: u8) -> HybridKeyMaterial {
    let fill = |n: usize, tag: u8| -> Vec<u8> {
        (0..n)
            .map(|i| seed.wrapping_add(tag).wrapping_add(i as u8))
            .collect()
    };
    HybridKeyMaterial {
        ed25519_signing_raw: fill(ED25519_PUBKEY_BYTES, 0x01),
        mldsa65_signing_raw: fill(MLDSA65_VERKEY_BYTES, 0x02),
        x25519_agreement_raw: fill(X25519_PUBKEY_BYTES, 0x03),
        mlkem768_encap_raw: fill(MLKEM768_ENCAP_BYTES, 0x04),
        next_ed25519_signing_raw: fill(ED25519_PUBKEY_BYTES, 0x11),
        next_mldsa65_signing_raw: fill(MLDSA65_VERKEY_BYTES, 0x12),
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct CesrKeys {
    pub ed25519_signing: String,
    pub mldsa65_signing: String,
    pub x25519_agreement: String,
    pub mlkem768_encap: String,
    pub next_ed25519_digest: String,
    pub next_mldsa65_digest: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct HybridInceptionResult {
    pub aid: String,
    pub said: String,
    pub inception_event: Value,
    pub raw_bytes_b64: String,
    pub cipher_suite: String,
    pub cesr: CesrKeys,
    pub public_key: String,
    pub next_key_digest: String,
}

fn material_to_cesr(m: &HybridKeyMaterial) -> Result<CesrKeys, String> {
    Ok(CesrKeys {
        ed25519_signing: ed25519_verfer_qb64(&m.ed25519_signing_raw)?,
        mldsa65_signing: encode_large_fixed(
            CESR_MLDSA65_VERKEY,
            &m.mldsa65_signing_raw,
            MLDSA65_VERKEY_BYTES,
        )?,
        x25519_agreement: encode_large_fixed(
            CESR_X25519_PUBKEY,
            &m.x25519_agreement_raw,
            X25519_PUBKEY_BYTES,
        )?,
        mlkem768_encap: encode_large_fixed(
            CESR_MLKEM768_ENCAP,
            &m.mlkem768_encap_raw,
            MLKEM768_ENCAP_BYTES,
        )?,
        next_ed25519_digest: blake3_qb64(&m.next_ed25519_signing_raw)?,
        next_mldsa65_digest: blake3_qb64(&m.next_mldsa65_signing_raw)?,
    })
}

pub fn build_hybrid_inception(m: &HybridKeyMaterial) -> Result<HybridInceptionResult, String> {
    let cesr = material_to_cesr(m)?;
    let k = vec![cesr.ed25519_signing.clone(), cesr.mldsa65_signing.clone()];
    let n = vec![
        cesr.next_ed25519_digest.clone(),
        cesr.next_mldsa65_digest.clone(),
    ];
    let ka = vec![cesr.x25519_agreement.clone(), cesr.mlkem768_encap.clone()];
    let anchor = vec![AnchorSeal {
        ia: CIPHER_SUITE_IA_HYBRID_1,
        ka: &ka,
    }];

    let wire = IcpWire {
        v: String::new(),
        t: "icp",
        d: "",
        i: "",
        s: "0",
        kt: "1",
        k: &k,
        nt: "1",
        n: &n,
        bt: "0",
        b: &[],
        c: &[],
        a: &anchor,
    };

    let dummy = "#".repeat(SAID_DUMMY_LEN);
    let (raw, digest) = makify_icp_wire(&wire, &dummy)?;

    let inception_event = json!({
        "v": versify(raw.len()),
        "t": "icp",
        "d": digest,
        "i": digest,
        "s": "0",
        "kt": "1",
        "k": k,
        "nt": "1",
        "n": n,
        "bt": "0",
        "b": [],
        "c": [],
        "a": [{"ia": CIPHER_SUITE_IA_HYBRID_1, "ka": ka}],
    });

    Ok(HybridInceptionResult {
        aid: digest.clone(),
        said: digest,
        inception_event,
        raw_bytes_b64: base64::encode(&raw),
        cipher_suite: CIPHER_SUITE_IA_HYBRID_1.to_string(),
        cesr: cesr.clone(),
        public_key: cesr.ed25519_signing.clone(),
        next_key_digest: cesr.next_ed25519_digest.clone(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    include!("../../tests/pqc_golden_constants.rs");

    #[test]
    fn cross_engine_byte_identity_seed0() {
        let m = synthetic_hybrid_key_material(0);
        let r = build_hybrid_inception(&m).expect("build");
        assert_eq!(r.aid, EXPECTED_AID);
        assert_eq!(r.said, EXPECTED_SAID);
        assert_eq!(r.aid, r.said);
        assert_eq!(r.raw_bytes_b64.len(), EXPECTED_RAW_B64_LEN);
        if let Some(pinned) = load_pinned_raw_b64() {
            assert_eq!(r.raw_bytes_b64, pinned, "raw_bytes_b64 mismatch vs keripy golden");
        }
    }

    #[test]
    fn synthetic_hybrid_inception_structure() {
        let m = synthetic_hybrid_key_material(0);
        let res = build_hybrid_inception(&m).expect("build");
        assert_eq!(res.cipher_suite, CIPHER_SUITE_IA_HYBRID_1);
        assert!(!res.aid.is_empty());
        assert_eq!(res.aid, res.said);
        assert!(res.cesr.mldsa65_signing.starts_with(CESR_MLDSA65_VERKEY));
        assert!(res.cesr.mlkem768_encap.starts_with(CESR_MLKEM768_ENCAP));
    }
}

File: identity-agent-core/iacrypto/hybrid_inception.go
package iacrypto

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"fmt"
	keri "github.com/grapeid/keri-go"
)

// HybridKeyMaterial holds raw key bytes for hybrid inception (synthetic or caller-supplied).
type HybridKeyMaterial struct {
	Ed25519SigningRaw     []byte
	MLDSA65SigningRaw     []byte
	X25519AgreementRaw    []byte
	MLKEM768EncapRaw      []byte
	NextEd25519SigningRaw []byte
	NextMLDSA65SigningRaw []byte
}

// SyntheticHybridKeyMaterial returns deterministic synthetic bytes for harness vectors.
func SyntheticHybridKeyMaterial(seed int) HybridKeyMaterial {
	fill := func(n, tag int) []byte {
		out := make([]byte, n)
		for i := range out {
			out[i] = byte((seed + tag + i) % 256)
		}
		return out
	}
	return HybridKeyMaterial{
		Ed25519SigningRaw:     fill(Ed25519PubkeyBytes, 0x01),
		MLDSA65SigningRaw:     fill(MLDSA65VerkeyBytes, 0x02),
		X25519AgreementRaw:    fill(X25519PubkeyBytes, 0x03),
		MLKEM768EncapRaw:      fill(MLKEM768EncapBytes, 0x04),
		NextEd25519SigningRaw: fill(Ed25519PubkeyBytes, 0x11),
		NextMLDSA65SigningRaw: fill(MLDSA65VerkeyBytes, 0x12),
	}
}

type cesrKeys struct {
	Ed25519Signing    string `json:"ed25519_signing"`
	MLDSA65Signing    string `json:"mldsa65_signing"`
	X25519Agreement   string `json:"x25519_agreement"`
	MLKEM768Encap     string `json:"mlkem768_encap"`
	NextEd25519Digest string `json:"next_ed25519_digest"`
	NextMLDSA65Digest string `json:"next_mldsa65_digest"`
}

// HybridInceptionResult mirrors the keripy driver response shape.
type HybridInceptionResult struct {
	AID            string                 `json:"aid"`
	SAID           string                 `json:"said"`
	InceptionEvent map[string]interface{} `json:"inception_event"`
	RawBytesB64    string                 `json:"raw_bytes_b64"`
	CipherSuite    string                 `json:"cipher_suite"`
	CESR           cesrKeys               `json:"cesr"`
	// Delegator is the identity this one acts under, empty when it acts alone.
	Delegator     string `json:"delegator,omitempty"`
	PublicKey     string `json:"public_key"`
	NextKeyDigest string `json:"next_key_digest"`
}

func ed25519VerferQB64(raw32 []byte) (string, error) {
	if len(raw32) != Ed25519PubkeyBytes {
		return "", fmt.Errorf("Ed25519 public key must be 32 bytes")
	}
	_ = ed25519.PublicKey(raw32)
	return MatterFixedQB64("D", raw32)
}

func materialToCESR(m HybridKeyMaterial) (cesrKeys, error) {
	ed, err := ed25519VerferQB64(m.Ed25519SigningRaw)
	if err != nil {
		return cesrKeys{}, err
	}
	mldsa, err := EncodeLargeFixed(CESRMLDSA65Verkey, m.MLDSA65SigningRaw, MLDSA65VerkeyBytes)
	if err != nil {
		return cesrKeys{}, err
	}
	x, err := EncodeLargeFixed(CESRX25519Pubkey, m.X25519AgreementRaw, X25519PubkeyBytes)
	if err != nil {
		return cesrKeys{}, err
	}
	kem, err := EncodeLargeFixed(CESRMLKEM768Encap, m.MLKEM768EncapRaw, MLKEM768EncapBytes)
	if err != nil {
		return cesrKeys{}, err
	}
	nEd, err := Blake3QB64(m.NextEd25519SigningRaw)
	if err != nil {
		return cesrKeys{}, err
	}
	nMldsa, err := Blake3QB64(m.NextMLDSA65SigningRaw)
	if err != nil {
		return cesrKeys{}, err
	}
	return cesrKeys{
		Ed25519Signing:    ed,
		MLDSA65Signing:    mldsa,
		X25519Agreement:   x,
		MLKEM768Encap:     kem,
		NextEd25519Digest: nEd,
		NextMLDSA65Digest: nMldsa,
	}, nil
}

// BuildHybridInception constructs keri 1.1.17 conformant hybrid icp (SerderKERI makify).
func BuildHybridInception(m HybridKeyMaterial) (*HybridInceptionResult, error) {
	return buildHybrid(m, "")
}

// BuildHybridDelegatedInception is the same event for an identity that acts
// under another one's authority, naming its delegator.
//
// This is what a sealed machine gets. It generates these keys inside the
// enclave and they never leave it, so it holds the private half it needs in
// order to decrypt — while the owner's root seed stays where it is and never
// reaches the hardware. The owner's authority arrives as a signature over this
// event rather than as key material, which is the whole reason the machine can
// have an identity at all without being given the owner's.
//
// The encryption keys are anchored in THIS event, so the machine's own
// identifier commits to them: substituting them means substituting the
// identifier, which the delegation no longer covers.
func BuildHybridDelegatedInception(m HybridKeyMaterial, delegatorAID string) (*HybridInceptionResult, error) {
	if delegatorAID == "" {
		return nil, fmt.Errorf("a delegated inception must name its delegator, or nothing establishes who this identity acts for")
	}
	return buildHybrid(m, delegatorAID)
}

func buildHybrid(m HybridKeyMaterial, delegatorAID string) (*HybridInceptionResult, error) {
	cesr, err := materialToCESR(m)
	if err != nil {
		return nil, err
	}

	// Built with keri-go rather than this package's own serialiser, so there is
	// one KERI implementation in this codebase instead of two. The two were
	// compared byte-for-byte on this exact event before the swap — see
	// kerigo_agreement_test.go, which stays as the guard against them drifting
	// apart again.
	anchor, err := json.Marshal(map[string]any{
		"ia": CipherSuiteIAHybrid1,
		"ka": []string{cesr.X25519Agreement, cesr.MLKEM768Encap},
	})
	if err != nil {
		return nil, err
	}
	// A delegated inception is a different event type carrying the delegator,
	// not an ordinary one with a field added — the identifier derives from the
	// whole event, so the two are different identities.
	var raw []byte
	if delegatorAID != "" {
		raw, err = keri.BuildDelegatedInception(keri.DelegationInput{
			Keys:        []string{cesr.Ed25519Signing, cesr.MLDSA65Signing},
			Isith:       json.RawMessage(`"1"`),
			NextDigests: []string{cesr.NextEd25519Digest, cesr.NextMLDSA65Digest},
			Nsith:       json.RawMessage(`"1"`),
			Data:        []json.RawMessage{anchor},
			Delegator:   delegatorAID,
		})
	} else {
		raw, err = keri.BuildInception(keri.InceptionInput{
			Keys:        []string{cesr.Ed25519Signing, cesr.MLDSA65Signing},
			Isith:       json.RawMessage(`"1"`),
			NextDigests: []string{cesr.NextEd25519Digest, cesr.NextMLDSA65Digest},
			Nsith:       json.RawMessage(`"1"`),
			Data:        []json.RawMessage{anchor},
		})
	}
	if err != nil {
		return nil, err
	}
	ev, err := keri.ParseEvent(raw)
	if err != nil {
		return nil, err
	}
	var event map[string]interface{}
	if err := json.Unmarshal(raw, &event); err != nil {
		return nil, err
	}

	return &HybridInceptionResult{
		AID:            ev.Identifier,
		SAID:           ev.SAID,
		InceptionEvent: event,
		Delegator:      delegatorAID,
		RawBytesB64:    base64.StdEncoding.EncodeToString(raw),
		CipherSuite:    CipherSuiteIAHybrid1,
		CESR:           cesr,
		PublicKey:      cesr.Ed25519Signing,
		NextKeyDigest:  cesr.NextEd25519Digest,
	}, nil
}
